File: baekjoon/10800.py
import sys

input = sys.stdin.readline

n = int(input())

balls = []
answer = dict()
ball_sum_dict = dict()

for i in range(n):
    c, s = map(int, input().split())
    balls.append((s, c, i))
    ball_sum_dict[c] = 0
    answer[i] = 0

balls.sort()
total = 0
j = 0
for i in range(n):
    while balls[j][0] < balls[i][0]:
        print("i, j", i, j)
        total += balls[j][0]
        ball_sum_dict[balls[j][1]] += balls[j][0]
        j += 1
    
    answer[balls[i][2]] = total - ball_sum_dict[balls[i][1]]

print(ball_sum_dict)
print(balls)
print(answer)

File: java/src/baekjoon/boj2022/_10800.java
package baekjoon.boj2022;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.StringTokenizer;

public class _10800 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        ArrayList<Ball> balls = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int c = Integer.parseInt(st.nextToken());
            int s = Integer.parseInt(st.nextToken());
            balls.add(new Ball(i, c, s));
        }
        Collections.sort(balls, new Comparator<Ball>() {
            @Override
            public int compare(Ball o1, Ball o2) {
                return o1.size - o2.size;
            }
        });

        int[] answer = new int[n];
        int[] color = new int[n + 1];
        int j = 0;
        int total = 0;

        for (int i = 0; i < n; i++) {
            Ball a = balls.get(i);
            Ball b = balls.get(j);

            while (b.size < a.size) {
                total += b.size;
                color[b.color] += b.size;

                b = balls.get(++j);
            }

            answer[a.index] = total - color[a.color];
        }

        StringBuffer sb = new StringBuffer();
        for (int i : answer) {
            sb.append(i).append("\n");
        }

        System.out.println(sb);

    }

    private static class Ball {
        int index;
        int color;
        int size;

        public Ball(int index, int color, int size) {
            this.index = index;
            this.color = color;
            this.size = size;
        }
    }
}
